In [205]:
import warnings
warnings.filterwarnings('ignore')

**About Book Crossing Dataset**<br>

This dataset has been compiled by Cai-Nicolas Ziegler in 2004, and it comprises of three tables for users, books and ratings. Explicit ratings are expressed on a scale from 1-10 (higher values denoting higher appreciation) and implicit rating is expressed by 0.

Reference: http://www2.informatik.uni-freiburg.de/~cziegler/BX/ 

**Objective**

This project entails building a Book Recommender System for users based on user-based and item-based collaborative filtering approaches.

#### Execute the below cell to load the datasets

In [206]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as pls
import seaborn as sns 

In [207]:
#Loading data
books = pd.read_csv("books.csv", sep=";", error_bad_lines=False, encoding="latin-1")
books.columns = ['ISBN', 'bookTitle', 'bookAuthor', 'yearOfPublication', 'publisher', 'imageUrlS', 'imageUrlM', 'imageUrlL']

users = pd.read_csv('users.csv', sep=';', error_bad_lines=False, encoding="latin-1")
users.columns = ['userID', 'Location', 'Age']

ratings = pd.read_csv('ratings.csv', sep=';', error_bad_lines=False, encoding="latin-1")
ratings.columns = ['userID', 'ISBN', 'bookRating']

b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkipping line 185738: expected 8 fields, saw 9\n'
b'Skipping line 209388: expected 8 fields, saw 9\nSkipping line 220626: expected 8 fields, saw 9\nSkipping line 227933: expected 8 fields, saw 11\nSkipping line 228957: expected 8 fields, saw 10\nSkipping line 245933: expected 8 fields, saw 9\nSkipping line 251296: expected 8 fields, saw 9\nSkipping line 259941: expected 8 fields, saw 9\nSkipping line 261529: expected 8 fields, saw 9\n'


### Check no.of records and features given in each dataset

In [208]:
print ('shape of books data set record ',books.shape[0])
print ('shape of books data set features ',books.shape[1])

shape of books data set record  271360
shape of books data set features  8


In [209]:
print ('shape of users data set record ',users.shape[0])
print ('shape of users data set features ',users.shape[1])

shape of users data set record  278858
shape of users data set features  3


In [210]:
print ('shape of ratings data set record ',ratings.shape[0])
print ('shape of ratings data set features ',ratings.shape[1])

shape of ratings data set record  1149780
shape of ratings data set features  3


## Exploring books dataset

In [211]:
books.head()

ISBN                                          bookTitle  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

             bookAuthor yearOfPublication                   publisher  \
0    Mark P. O. Morford              2002     Oxford University Press   
1  Richard Bruce Wright              2001       HarperFlamingo Canada   
2          Carlo D'Este              1991             HarperPerennial   
3      Gina Bari Kolata              1999        Farrar Straus Giroux   
4       E. J. W. Barber              1999  W. W. Norton &amp; Company   

                                           imageUrlS  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                           imageUrlM  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                           imageUrlL  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

### Drop last three columns containing image URLs which will not be required for analysis

In [212]:
books = books.drop(columns= ['imageUrlS','imageUrlM','imageUrlL'],axis=1)

In [213]:
books.head()

ISBN                                          bookTitle  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

             bookAuthor yearOfPublication                   publisher  
0    Mark P. O. Morford              2002     Oxford University Press  
1  Richard Bruce Wright              2001       HarperFlamingo Canada  
2          Carlo D'Este              1991             HarperPerennial  
3      Gina Bari Kolata              1999        Farrar Straus Giroux  
4       E. J. W. Barber              1999  W. W. Norton &amp; Company

**yearOfPublication**

### Check unique values of yearOfPublication


In [214]:
books['yearOfPublication'].unique()

array([2002, 2001, 1991, 1999, 2000, 1993, 1996, 1988, 2004, 1998, 1994,
       2003, 1997, 1983, 1979, 1995, 1982, 1985, 1992, 1986, 1978, 1980,
       1952, 1987, 1990, 1981, 1989, 1984, 0, 1968, 1961, 1958, 1974,
       1976, 1971, 1977, 1975, 1965, 1941, 1970, 1962, 1973, 1972, 1960,
       1966, 1920, 1956, 1959, 1953, 1951, 1942, 1963, 1964, 1969, 1954,
       1950, 1967, 2005, 1957, 1940, 1937, 1955, 1946, 1936, 1930, 2011,
       1925, 1948, 1943, 1947, 1945, 1923, 2020, 1939, 1926, 1938, 2030,
       1911, 1904, 1949, 1932, 1928, 1929, 1927, 1931, 1914, 2050, 1934,
       1910, 1933, 1902, 1924, 1921, 1900, 2038, 2026, 1944, 1917, 1901,
       2010, 1908, 1906, 1935, 1806, 2021, '2000', '1995', '1999', '2004',
       '2003', '1990', '1994', '1986', '1989', '2002', '1981', '1993',
       '1983', '1982', '1976', '1991', '1977', '1998', '1992', '1996',
       '0', '1997', '2001', '1974', '1968', '1987', '1984', '1988',
       '1963', '1956', '1970', '1985', '1978', '1973', '1980'

In [215]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 5 columns):
ISBN                 271360 non-null object
bookTitle            271360 non-null object
bookAuthor           271359 non-null object
yearOfPublication    271360 non-null object
publisher            271358 non-null object
dtypes: object(5)
memory usage: 10.4+ MB


As it can be seen from above that there are some incorrect entries in this field. It looks like Publisher names 'DK Publishing Inc' and 'Gallimard' have been incorrectly loaded as yearOfPublication in dataset due to some errors in csv file.


Also some of the entries are strings and same years have been entered as numbers in some places. We will try to fix these things in the coming questions.

### Check the rows having 'DK Publishing Inc' as yearOfPublication

In [217]:
yr_bad_entries = books[books['yearOfPublication'].str.contains("DK Publishing Inc")== True]
yr_bad_entries

ISBN                                          bookTitle  \
209538  078946697X  DK Readers: Creating the X-Men, How It All Beg...   
221678  0789466953  DK Readers: Creating the X-Men, How Comic Book...   

       bookAuthor  yearOfPublication  \
209538       2000  DK Publishing Inc   
221678       2000  DK Publishing Inc   

                                                publisher  
209538  http://images.amazon.com/images/P/078946697X.0...  
221678  http://images.amazon.com/images/P/0789466953.0...

### Drop the rows having `'DK Publishing Inc'` and `'Gallimard'` as `yearOfPublication`

In [218]:
books = books[~books['yearOfPublication'].str.contains("Gallimard",na=False)]
books = books[~books['yearOfPublication'].str.contains("DK Publishing Inc",na=False)]
books

ISBN                                          bookTitle  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271355  0440400988                         There's a Bat in Bunk Five   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                  bookAuthor yearOfPublication  \
0         Mark P. O. Morford              2002   
1       Richard Bruce Wright              2001   
2               Carlo D'Este              1991   
3           Gina Bari Kolata              1999   
4            E. J. W. Barber              1999   
...                      ...               ...   
271355        Paula Danziger              1988   
271356            Teri Sloat              1991   
271357      Christine Wicker              2004   
271358                 Plato              1996   
271359   Christopher  Biffle              2000   

                                               publisher  
0                                Oxford University Press  
1                                  HarperFlamingo Canada  
2                                        HarperPerennial  
3                                   Farrar Straus Giroux  
4                             W. W. Norton &amp; Company  
...                                                  ...  
271355                   Random House Childrens Pub (Mm)  
271356                                      Dutton Books  
271357                                HarperSanFrancisco  
271358                           Oxford University Press  
271359  McGraw-Hill Humanities/Social Sciences/Languages  

[271357 rows x 5 columns]

>- We Can observe rows reduce by three i.e ("271360 to 271357") after droping strings from yearofpublication

### Change the datatype of yearOfPublication to 'int'

In [219]:
books['yearOfPublication'] = books['yearOfPublication'].astype(int)

In [220]:
books.dtypes

ISBN                 object
bookTitle            object
bookAuthor           object
yearOfPublication     int32
publisher            object
dtype: object

### Drop NaNs in `'publisher'` column


In [221]:
books  = books[~books['publisher'].isna()]
books.shape

(271355, 5)

## Exploring Users dataset

In [222]:
print(users.shape)
users.head()

(278858, 3)


userID                            Location   Age
0       1                  nyc, new york, usa   NaN
1       2           stockton, california, usa  18.0
2       3     moscow, yukon territory, russia   NaN
3       4           porto, v.n.gaia, portugal  17.0
4       5  farnborough, hants, united kingdom   NaN

### Get all unique values in ascending order for column `Age`

In [223]:
users['Age'].unique()

array([ nan,  18.,  17.,  61.,  26.,  14.,  25.,  19.,  46.,  55.,  32.,
        24.,  20.,  34.,  23.,  51.,  31.,  21.,  44.,  30.,  57.,  43.,
        37.,  41.,  54.,  42.,  50.,  39.,  53.,  47.,  36.,  28.,  35.,
        13.,  58.,  49.,  38.,  45.,  62.,  63.,  27.,  33.,  29.,  66.,
        40.,  15.,  60.,   0.,  79.,  22.,  16.,  65.,  59.,  48.,  72.,
        56.,  67.,   1.,  80.,  52.,  69.,  71.,  73.,  78.,   9.,  64.,
       103., 104.,  12.,  74.,  75., 231.,   3.,  76.,  83.,  68., 119.,
        11.,  77.,   2.,  70.,  93.,   8.,   7.,   4.,  81., 114., 230.,
       239.,  10.,   5., 148., 151.,   6., 101., 201.,  96.,  84.,  82.,
        90., 123., 244., 133.,  91., 128.,  94.,  85., 141., 110.,  97.,
       219.,  86., 124.,  92., 175., 172., 209., 212., 237.,  87., 162.,
       100., 156., 136.,  95.,  89., 106.,  99., 108., 210.,  88., 199.,
       147., 168., 132., 159., 186., 152., 102., 116., 200., 115., 226.,
       137., 207., 229., 138., 109., 105., 228., 18

Age column has some invalid entries like nan, 0 and very high values like 100 and above

### Values below 5 and above 90 do not make much sense for our book rating case...hence replace these by NaNs

In [224]:
#users.loc[(users.Age>90) |(users.Age<5)]=np.NaN
users['Age'] = users['Age'].apply(lambda x:np.nan if x > 90 else x)
users['Age'] = users['Age'].apply(lambda x:np.nan if x < 5 else x)
users['Age'].mean()

34.72384041634689

### Replace null values in column `Age` with mean

In [225]:
users['Age'] = users['Age'].replace(np.nan,users['Age'].mean())
users['Age'].isnull().sum()

0

### Change the datatype of `Age` to `int`

In [226]:
users['Age'] = users['Age'].astype(int)


In [227]:
print(sorted(users.Age.unique()))

[5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90]


## Exploring the Ratings Dataset

### check the shape

In [228]:
ratings.shape

(1149780, 3)

In [229]:
n_users = users.shape[0]
n_books = books.shape[0]

In [230]:
ratings.head(5)

userID        ISBN  bookRating
0  276725  034545104X           0
1  276726  0155061224           5
2  276727  0446520802           0
3  276729  052165615X           3
4  276729  0521795028           6

### Ratings dataset should have books only which exist in our books dataset. Drop the remaining rows

In [231]:
ratings = ratings[ratings.ISBN.isin(books.ISBN)]
ratings.shape

(1031130, 3)

### Ratings dataset should have ratings from users which exist in users dataset. Drop the remaining rows

In [232]:
ratings = ratings[ratings.userID.isin(users.userID)]
ratings.shape 

(1031130, 3)

### Consider only ratings from 1-10 and leave 0s in column `bookRating`

In [233]:
ratings = ratings[ratings['bookRating']>=1]
ratings = ratings[ratings['bookRating']<=10]
ratings.shape

(383839, 3)

### Find out which rating has been given highest number of times

In [234]:
ratings['bookRating'].value_counts()

8     91804
10    71225
7     66401
9     60776
5     45355
6     31687
4      7617
3      5118
2      2375
1      1481
Name: bookRating, dtype: int64

>- From the above count values we can see rating 8 has been given highest number of times

### **Collaborative Filtering Based Recommendation Systems**

### For more accurate results only consider users who have rated atleast 100 books

In [273]:
#pd.crosstab(ratings['userID'],ratings['bookRating'])

In [236]:
more_100 = pd.DataFrame(ratings['userID'].value_counts() > 99)
more = more_100[more_100['userID'] == True].index
more # more contains all the userID who have rated atleast 100 books 

Int64Index([ 11676,  98391, 189835, 153662,  23902, 235105,  76499, 171118,
             16795, 248718,
            ...
            117384,  36299, 169682, 211919, 156300,  95010,  33145,  26544,
            208406,  36609],
           dtype='int64', length=449)

In [237]:
(ratings['userID'].value_counts() > 99 ).sum()

449

In [238]:
ratings = ratings[ratings['userID'].isin(more)]
ratings

userID        ISBN  bookRating
1456     277427  002542730X          10
1458     277427  003008685X           8
1461     277427  0060006641          10
1465     277427  0060542128           7
1474     277427  0061009059           9
...         ...         ...         ...
1147587  275970  185649814X           7
1147592  275970  1860462588           8
1147599  275970  1886411077           6
1147611  275970  3411086211          10
1147615  275970  9626340762           8

[103269 rows x 3 columns]

In [239]:
len(ratings['userID'].value_counts()>99) # conforming for filltering 

449

In [274]:
book_data1 = pd.merge(ratings, books, on='ISBN')
book_data1.head()

userID        ISBN  bookRating  \
0  277427  002542730X          10   
1   11676  002542730X           6   
2   12538  002542730X          10   
3   52584  002542730X          10   
4  110934  002542730X           6   

                                           bookTitle         bookAuthor  \
0  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
1  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
2  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
3  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
4  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   

   yearOfPublication                  publisher  
0               1994  John Wiley &amp; Sons Inc  
1               1994  John Wiley &amp; Sons Inc  
2               1994  John Wiley &amp; Sons Inc  
3               1994  John Wiley &amp; Sons Inc  
4               1994  John Wiley &amp; Sons Inc

### Generating ratings matrix from explicit ratings


#### Note: since NaNs cannot be handled by training algorithms, replace these by 0, which indicates absence of ratings

In [241]:
"""
from sklearn.model_selection import train_test_split

trainDF, tempDF = train_test_split(ratings, test_size = 0.2, random_state = 100)
"""


'\nfrom sklearn.model_selection import train_test_split\n\ntrainDF, tempDF = train_test_split(ratings, test_size = 0.2, random_state = 100)\n'

In [242]:
"""
testDF = tempDF.copy()
testDF.head()
testDF.shape
trainDF.shape 
"""

'\ntestDF = tempDF.copy()\ntestDF.head()\ntestDF.shape\ntrainDF.shape \n'

In [243]:
"""
ratings = pd.concat([trainDF, tempDF])
print (ratings.shape)
ratings.head()
"""

'\nratings = pd.concat([trainDF, tempDF])\nprint (ratings.shape)\nratings.head()\n'

In [244]:
R_df = ratings.pivot(index = 'userID', columns = 'ISBN', values = 'bookRating').fillna(0)
R_df.head()
R_df.shape

(449, 66572)

In [245]:
R_df.isna().sum().sum()

0

### Generate the predicted ratings using SVD with no.of singular values to be 50

In [246]:
from scipy.sparse.linalg import svds
U, sigma, Vt = svds(R_df, k = 50)

In [247]:
sigma

array([147.92121613, 149.3438051 , 150.07400599, 152.20116297,
       152.87416391, 154.61308307, 154.80093432, 155.95760177,
       158.05646578, 159.21079484, 159.81670657, 162.01963916,
       162.77851768, 163.33054635, 166.02489324, 166.8162391 ,
       168.04972004, 170.77485167, 171.01325686, 173.29428498,
       174.57624968, 176.65724713, 178.61913749, 180.29517222,
       182.25079063, 184.10706957, 187.61687534, 189.75276623,
       190.96966388, 195.14643609, 199.83133018, 201.70083339,
       202.18713912, 203.48697581, 207.26449173, 209.92986988,
       213.23598777, 216.88280493, 224.26954726, 231.66186197,
       235.67095629, 249.9581775 , 252.02866425, 261.24756904,
       267.98197504, 281.0120779 , 293.69539562, 379.58327277,
       634.72875357, 680.30978318])

### Take a particular user_id

### Lets find the recommendations for user with id `2110`

#### Note: Execute the below cells to get the variables loaded

In [248]:
userID = 2110

In [249]:
user_id = 2 #2nd row in ratings matrix and predicted matrix

### Get the predicted ratings for userID `2110` and sort them in descending order

In [250]:
sigma = np.diag(sigma)

In [251]:
sigma

array([[147.92121613,   0.        ,   0.        , ...,   0.        ,
          0.        ,   0.        ],
       [  0.        , 149.3438051 ,   0.        , ...,   0.        ,
          0.        ,   0.        ],
       [  0.        ,   0.        , 150.07400599, ...,   0.        ,
          0.        ,   0.        ],
       ...,
       [  0.        ,   0.        ,   0.        , ..., 379.58327277,
          0.        ,   0.        ],
       [  0.        ,   0.        ,   0.        , ...,   0.        ,
        634.72875357,   0.        ],
       [  0.        ,   0.        ,   0.        , ...,   0.        ,
          0.        , 680.30978318]])

In [252]:
all_users_predicted_ratings = np.dot(np.dot(U, sigma), Vt)

In [275]:
preds_df = pd.DataFrame(all_users_predicted_ratings, columns = R_df.columns)
preds_df.head()

ISBN  0000913154  0001046438  000104687X  0001047213  0001047973  000104799X  \
0       0.025341   -0.002146   -0.001431   -0.002146   -0.002146    0.002971   
1      -0.010012   -0.003669   -0.002446   -0.003669   -0.003669    0.001075   
2      -0.015054   -0.015457   -0.010304   -0.015457   -0.015457    0.007281   
3      -0.021499    0.035602    0.023735    0.035602    0.035602    0.030307   
4       0.002077   -0.007965   -0.005310   -0.007965   -0.007965    0.002947   

ISBN  0001048082  0001053736  0001053744  0001055607  ...  B000092Q0A  \
0      -0.003920    0.007035    0.007035    0.012316  ...    0.000180   
1       0.001440   -0.003500   -0.003500    0.001612  ...   -0.000363   
2      -0.014033    0.011941    0.011941    0.011796  ...   -0.000455   
3       0.024215   -0.001053   -0.001053    0.067579  ...    0.002971   
4       0.003057    0.000231    0.000231    0.006080  ...    0.002120   

ISBN  B00009EF82  B00009NDAN  B0000DYXID  B0000T6KHI  B0000VZEJQ  B0000X8HIE  \
0       0.000226    0.042081   -0.016804   -0.080028    0.004746    0.028314   
1       0.000403    0.008142    0.001104   -0.029224    0.000999    0.002363   
2       0.001907    0.047982    0.005737    0.117859    0.006945    0.003119   
3       0.009912    0.086248   -0.008818    0.016154    0.028848   -0.000125   
4       0.001597   -0.012181    0.009420    0.673459    0.002591   -0.008229   

ISBN  B00013AX9E  B0001I1KOG  B000234N3A  
0       0.000120   -0.001693    0.067503  
1      -0.000242    0.000029   -0.013059  
2      -0.000304    0.009009   -0.057692  
3       0.001981    0.031201   -0.046664  
4       0.001413    0.004918    0.047773  

[5 rows x 66572 columns]

In [255]:
def recommend_books(predictions_df, userID,userId, books_df, original_ratings_df, num_recommendations = False):
    user_row_number = userID   #UserID starts at zero not 1
    sorted_user_predictions = predictions_df.loc[user_row_number].sort_values(ascending = False)
    
    user_data = original_ratings_df[original_ratings_df.userID == (userId)]
    user_full = (user_data.merge(books, how = 'left', left_on = 'ISBN', right_on = 'ISBN').
                sort_values(['bookRating'], ascending = False)
                )
    print('User {0} has already rated {1} books.'.format(userID, user_full.dropna().shape[0]))
    print('Recommending the highest {0} predicted ratings books not already read.'.format(num_recommendations))
    
    recommendations = (books_df[~books_df['ISBN'].isin(user_full['ISBN'])].
                      merge(pd.DataFrame(sorted_user_predictions).reset_index(), how = 'left',
                           left_on = 'ISBN',
                           right_on = 'ISBN').
                      rename(columns = {user_row_number: 'Predictions'}).
                      sort_values('Predictions', ascending = False).
                      iloc[:num_recommendations, :-1])
    return user_full, recommendations, sorted_user_predictions, user_data, user_full
#R_df = rating.pivot(index = 'userID', columns = 'ISBN', values = 'bookRating').fillna(0)

In [256]:

already_rated, predictions, sorted_user_predictions, user_data, user_full = recommend_books(preds_df, 2,2110,books, ratings, 10)

User 2 has already rated 103 books.
Recommending the highest 10 predicted ratings books not already read.


### Already rated by user 

In [257]:
already_rated

userID        ISBN  bookRating  \
76    2110  067166865X          10   
52    2110  0590109715          10   
64    2110  0590629786          10   
63    2110  0590629778          10   
61    2110  059046678X          10   
..     ...         ...         ...   
49    2110  0515134384           5   
19    2110  037361490X           5   
4     2110  0151008116           5   
50    2110  0515136557           3   
84    2110  0688047211           1   

                                            bookTitle            bookAuthor  \
76  STAR TREK YESTERDAY'S SON (Star Trek: The Orig...          A.C. Crispin   
52  The Andalite Chronicles (Elfangor's Journey, A...   Katherine Applegate   
64                      The Visitor (Animorphs, No 2)       K. A. Applegate   
63                     The Invasion (Animorphs, No 1)       K. A. Applegate   
61                                       The Yearbook        Peter Lerangis   
..                                                ...                   ...   
49                      The Cat Who Went Up the Creek  Lilian Jackson Braun   
19                      Age of War  (Super Bolan #90)         Don Pendleton   
4                                          Life of Pi           Yann Martel   
50                 The Cat Who Brought Down the House  Lilian Jackson Braun   
84                           The Lives of John Lennon        Albert Goldman   

    yearOfPublication      publisher  
76               1988     Audioworks  
52               1997          Apple  
64               1996     Scholastic  
63               1996     Scholastic  
61               1994     Scholastic  
..                ...            ...  
49               2003     Jove Books  
19               2003     Gold Eagle  
4                2002       Harcourt  
50               2004     Jove Books  
84               1988  Harpercollins  

[103 rows x 7 columns]

### reccendation for User

In [258]:
predictions

ISBN                                          bookTitle  \
407    0316666343                          The Lovely Bones: A Novel   
2116   0345350499                                The Mists of Avalon   
2438   0440214041                                  The Pelican Brief   
455    044021145X                                           The Firm   
521    0312195516                The Red Tent (Bestselling Backlist)   
20670  0345318862      Golem in the Gears (Xanth Novels (Paperback))   
4810   0345313151  Bearing an Hourglass (Incarnations of Immortal...   
6320   0380752891           Man from Mundania (Xanth Trilogy, No 12)   
44448  051511605X                                    Undue Influence   
8977   043936213X     Harry Potter and the Sorcerer's Stone (Book 1)   

                  bookAuthor  yearOfPublication                      publisher  
407             Alice Sebold               2002                  Little, Brown  
2116   MARION ZIMMER BRADLEY               1987                        Del Rey  
2438            John Grisham               1993                           Dell  
455             John Grisham               1992   Bantam Dell Publishing Group  
521            Anita Diamant               1998                    Picador USA  
20670          PIERS ANTHONY               1986                        Del Rey  
4810           Piers Anthony               1991                  Del Rey Books  
6320           Piers Anthony               1990  Harper Mass Market Paperbacks  
44448    Steven Paul Martini               1995                     Jove Books  
8977           J. K. Rowling               2001                     Scholastic

In [259]:
#user_records[220]

In [260]:
sorted_user_predictions

ISBN
0316666343    1.015397
059035342X    0.778665
0345350499    0.697309
0440214041    0.665439
044021145X    0.663549
                ...   
0380709562   -0.162346
0553213164   -0.173301
0440219078   -0.174497
0807508527   -0.204146
0743235150   -0.209452
Name: 2, Length: 66572, dtype: float64

### User Data 

In [261]:
user_data

userID        ISBN  bookRating
14448    2110  0060987529           7
14449    2110  0064472779           8
14450    2110  0140022651          10
14452    2110  0142302163           8
14453    2110  0151008116           5
...       ...         ...         ...
14603    2110  1558504184           8
14605    2110  1561008931           7
14606    2110  1565111575          10
14608    2110  1570420564          10
14609    2110  1572970944           6

[103 rows x 3 columns]

In [262]:
user_data.shape 

(103, 3)

In [263]:
user_full

userID        ISBN  bookRating  \
76    2110  067166865X          10   
52    2110  0590109715          10   
64    2110  0590629786          10   
63    2110  0590629778          10   
61    2110  059046678X          10   
..     ...         ...         ...   
49    2110  0515134384           5   
19    2110  037361490X           5   
4     2110  0151008116           5   
50    2110  0515136557           3   
84    2110  0688047211           1   

                                            bookTitle            bookAuthor  \
76  STAR TREK YESTERDAY'S SON (Star Trek: The Orig...          A.C. Crispin   
52  The Andalite Chronicles (Elfangor's Journey, A...   Katherine Applegate   
64                      The Visitor (Animorphs, No 2)       K. A. Applegate   
63                     The Invasion (Animorphs, No 1)       K. A. Applegate   
61                                       The Yearbook        Peter Lerangis   
..                                                ...                   ...   
49                      The Cat Who Went Up the Creek  Lilian Jackson Braun   
19                      Age of War  (Super Bolan #90)         Don Pendleton   
4                                          Life of Pi           Yann Martel   
50                 The Cat Who Brought Down the House  Lilian Jackson Braun   
84                           The Lives of John Lennon        Albert Goldman   

    yearOfPublication      publisher  
76               1988     Audioworks  
52               1997          Apple  
64               1996     Scholastic  
63               1996     Scholastic  
61               1994     Scholastic  
..                ...            ...  
49               2003     Jove Books  
19               2003     Gold Eagle  
4                2002       Harcourt  
50               2004     Jove Books  
84               1988  Harpercollins  

[103 rows x 7 columns]

### Create a dataframe with name `user_data` containing userID `2110` explicitly interacted books

In [264]:
user_data_2100 = ratings[ratings['userID'] == 2110]
user_data_2100.head()

userID        ISBN  bookRating
14448    2110  0060987529           7
14449    2110  0064472779           8
14450    2110  0140022651          10
14452    2110  0142302163           8
14453    2110  0151008116           5

In [265]:
user_data_2100.shape

(103, 3)

### Combine the user_data and and corresponding book data(`book_data`) in a single dataframe with name `user_full_info`

In [266]:
user_full_info = user_data_2100.merge(books,how = 'left' , left_on ='ISBN' , right_on = 'ISBN')
user_full_info=user_full_info.drop(columns=['userID','bookRating'],axis=1)

In [267]:
user_full_info.head()

ISBN                                    bookTitle       bookAuthor  \
0  0060987529  Confessions of an Ugly Stepsister : A Novel  Gregory Maguire   
1  0064472779                            All-American Girl        Meg Cabot   
2  0140022651           Journey to the Center of the Earth      Jules Verne   
3  0142302163                             The Ghost Sitter  Peni R. Griffin   
4  0151008116                                   Life of Pi      Yann Martel   

   yearOfPublication      publisher  
0               2000    Regan Books  
1               2003   HarperTrophy  
2               1965  Penguin Books  
3               2002   Puffin Books  
4               2002       Harcourt

### Get top 10 recommendations for above given userID from the books not already rated by that user

In [268]:
predictions

ISBN                                          bookTitle  \
407    0316666343                          The Lovely Bones: A Novel   
2116   0345350499                                The Mists of Avalon   
2438   0440214041                                  The Pelican Brief   
455    044021145X                                           The Firm   
521    0312195516                The Red Tent (Bestselling Backlist)   
20670  0345318862      Golem in the Gears (Xanth Novels (Paperback))   
4810   0345313151  Bearing an Hourglass (Incarnations of Immortal...   
6320   0380752891           Man from Mundania (Xanth Trilogy, No 12)   
44448  051511605X                                    Undue Influence   
8977   043936213X     Harry Potter and the Sorcerer's Stone (Book 1)   

                  bookAuthor  yearOfPublication                      publisher  
407             Alice Sebold               2002                  Little, Brown  
2116   MARION ZIMMER BRADLEY               1987                        Del Rey  
2438            John Grisham               1993                           Dell  
455             John Grisham               1992   Bantam Dell Publishing Group  
521            Anita Diamant               1998                    Picador USA  
20670          PIERS ANTHONY               1986                        Del Rey  
4810           Piers Anthony               1991                  Del Rey Books  
6320           Piers Anthony               1990  Harper Mass Market Paperbacks  
44448    Steven Paul Martini               1995                     Jove Books  
8977           J. K. Rowling               2001                     Scholastic

In [269]:
sorted_user_predictions1 = preds_df.loc[2].sort_values(ascending = False)

In [270]:
recommendations = (books[~books['ISBN'].isin(user_full['ISBN'])].
                      merge(pd.DataFrame(sorted_user_predictions1).reset_index(), how = 'left',
                           left_on = 'ISBN',
                           right_on = 'ISBN').rename(columns = {2: 'Predictions'}).
                      sort_values('Predictions', ascending = False).
                   iloc[:10, :-1])

In [271]:
recommendations

ISBN                                          bookTitle  \
407    0316666343                          The Lovely Bones: A Novel   
2116   0345350499                                The Mists of Avalon   
2438   0440214041                                  The Pelican Brief   
455    044021145X                                           The Firm   
521    0312195516                The Red Tent (Bestselling Backlist)   
20670  0345318862      Golem in the Gears (Xanth Novels (Paperback))   
4810   0345313151  Bearing an Hourglass (Incarnations of Immortal...   
6320   0380752891           Man from Mundania (Xanth Trilogy, No 12)   
44448  051511605X                                    Undue Influence   
8977   043936213X     Harry Potter and the Sorcerer's Stone (Book 1)   

                  bookAuthor  yearOfPublication                      publisher  
407             Alice Sebold               2002                  Little, Brown  
2116   MARION ZIMMER BRADLEY               1987                        Del Rey  
2438            John Grisham               1993                           Dell  
455             John Grisham               1992   Bantam Dell Publishing Group  
521            Anita Diamant               1998                    Picador USA  
20670          PIERS ANTHONY               1986                        Del Rey  
4810           Piers Anthony               1991                  Del Rey Books  
6320           Piers Anthony               1990  Harper Mass Market Paperbacks  
44448    Steven Paul Martini               1995                     Jove Books  
8977           J. K. Rowling               2001                     Scholastic